File: generated-code/python/codecraft/FileReadWrite/codegame/client_message.py
from codegame.debug_command import DebugCommand
from model.action import Action

class ClientMessage:
    """Message sent from client"""

    @staticmethod
    def read_from(stream):
        """Read ClientMessage from input stream
        """
        tag = stream.read_int()
        if tag == DebugMessage.TAG:
            return ClientMessage.DebugMessage.read_from(stream)
        if tag == ActionMessage.TAG:
            return ClientMessage.ActionMessage.read_from(stream)
        if tag == DebugUpdateDone.TAG:
            return ClientMessage.DebugUpdateDone.read_from(stream)
        if tag == RequestDebugState.TAG:
            return ClientMessage.RequestDebugState.read_from(stream)
        raise Exception("Unexpected tag value")

class DebugMessage(ClientMessage):
    """Ask app to perform new debug command"""

    TAG = 0

    def __init__(self, command):
        self.command = command
        """Command to perform"""

    @staticmethod
    def read_from(stream):
        """Read DebugMessage from input stream
        """
        command = DebugCommand.read_from(stream)
        return DebugMessage(command)

    def write_to(self, stream):
        """Write DebugMessage to output stream
        """
        stream.write_int(self.TAG)
        self.command.write_to(stream)

    def __repr__(self):
        return "DebugMessage(" + \
            repr(self.command) + \
            ")"

ClientMessage.DebugMessage = DebugMessage

class ActionMessage(ClientMessage):
    """Reply for ServerMessage::GetAction"""

    TAG = 1

    def __init__(self, action):
        self.action = action
        """Player's action"""

    @staticmethod
    def read_from(stream):
        """Read ActionMessage from input stream
        """
        action = Action.read_from(stream)
        return ActionMessage(action)

    def write_to(self, stream):
        """Write ActionMessage to output stream
        """
        stream.write_int(self.TAG)
        self.action.write_to(stream)

    def __repr__(self):
        return "ActionMessage(" + \
            repr(self.action) + \
            ")"

ClientMessage.ActionMessage = ActionMessage

class DebugUpdateDone(ClientMessage):
    """Signifies finish of the debug update"""

    TAG = 2

    def __init__(self):
        pass

    @staticmethod
    def read_from(stream):
        """Read DebugUpdateDone from input stream
        """
        return DebugUpdateDone()

    def write_to(self, stream):
        """Write DebugUpdateDone to output stream
        """
        stream.write_int(self.TAG)

    def __repr__(self):
        return "DebugUpdateDone(" + \
            ")"

ClientMessage.DebugUpdateDone = DebugUpdateDone

class RequestDebugState(ClientMessage):
    """Request debug state from the app"""

    TAG = 3

    def __init__(self):
        pass

    @staticmethod
    def read_from(stream):
        """Read RequestDebugState from input stream
        """
        return RequestDebugState()

    def write_to(self, stream):
        """Write RequestDebugState to output stream
        """
        stream.write_int(self.TAG)

    def __repr__(self):
        return "RequestDebugState(" + \
            ")"

ClientMessage.RequestDebugState = RequestDebugState

File: generated-code/pascal/codecraft/FileReadWrite/Codegame/ClientMessage.pas
unit ClientMessage;

interface

uses
    Action in 'Model/Action.pas',
    DebugCommand in 'Codegame/DebugCommand.pas',
    Stream,
    SysUtils;

type
    // Message sent from client
    TClientMessage = class
        // Write ClientMessage to output stream
        procedure WriteTo(stream: TStream); virtual; abstract;
        // Read ClientMessage from input stream
        class function ReadFrom(stream: TStream): TClientMessage; static;
    end;

type
    // Ask app to perform new debug command
    TClientMessageDebugMessage = class (TClientMessage)
        // Command to perform
        command: TDebugCommand;
        constructor Create(command: TDebugCommand);
        // Read ClientMessageDebugMessage from input stream
        class function ReadFrom(stream: TStream): TClientMessageDebugMessage; static;
        // Write ClientMessageDebugMessage to output stream
        procedure WriteTo(stream: TStream); override;
        function ToString: ansistring; override;
    end;

type
    // Reply for ServerMessage::GetAction
    TClientMessageActionMessage = class (TClientMessage)
        // Player's action
        action: TAction;
        constructor Create(action: TAction);
        // Read ClientMessageActionMessage from input stream
        class function ReadFrom(stream: TStream): TClientMessageActionMessage; static;
        // Write ClientMessageActionMessage to output stream
        procedure WriteTo(stream: TStream); override;
        function ToString: ansistring; override;
    end;

type
    // Signifies finish of the debug update
    TClientMessageDebugUpdateDone = class (TClientMessage)
        constructor Create();
        // Read ClientMessageDebugUpdateDone from input stream
        class function ReadFrom(stream: TStream): TClientMessageDebugUpdateDone; static;
        // Write ClientMessageDebugUpdateDone to output stream
        procedure WriteTo(stream: TStream); override;
        function ToString: ansistring; override;
    end;

type
    // Request debug state from the app
    TClientMessageRequestDebugState = class (TClientMessage)
        constructor Create();
        // Read ClientMessageRequestDebugState from input stream
        class function ReadFrom(stream: TStream): TClientMessageRequestDebugState; static;
        // Write ClientMessageRequestDebugState to output stream
        procedure WriteTo(stream: TStream); override;
        function ToString: ansistring; override;
    end;

implementation

class function TClientMessage.ReadFrom(stream: TStream): TClientMessage;
var tag: Int32;
begin
    tag := stream.ReadInt32;
    case tag of
        0: result := TClientMessageDebugMessage.ReadFrom(stream);
        1: result := TClientMessageActionMessage.ReadFrom(stream);
        2: result := TClientMessageDebugUpdateDone.ReadFrom(stream);
        3: result := TClientMessageRequestDebugState.ReadFrom(stream);
        else raise Exception.Create('Unexpected tag value');
    end;
end;

constructor TClientMessageDebugMessage.Create(command: TDebugCommand);
begin
    self.command := command;
end;

class function TClientMessageDebugMessage.ReadFrom(stream: TStream): TClientMessageDebugMessage;
var command: TDebugCommand;
begin
    command := TDebugCommand.ReadFrom(stream);
    result := TClientMessageDebugMessage.Create(command);
end;

procedure TClientMessageDebugMessage.WriteTo(stream: TStream);
begin
    stream.WriteInt32(0);
    command.WriteTo(stream);
end;

function TClientMessageDebugMessage.ToString: ansistring;
begin
    result := 'DebugMessage {';
    result += 'command=';
    result += command.ToString;
    result += '}';
end;

constructor TClientMessageActionMessage.Create(action: TAction);
begin
    self.action := action;
end;

class function TClientMessageActionMessage.ReadFrom(stream: TStream): TClientMessageActionMessage;
var action: TAction;
begin
    action := TAction.ReadFrom(stream);
    result := TClientMessageActionMessage.Create(action);
end;

procedure TClientMessageActionMessage.WriteTo(stream: TStream);
begin
    stream.WriteInt32(1);
    action.WriteTo(stream);
end;

function TClientMessageActionMessage.ToString: ansistring;
begin
    result := 'ActionMessage {';
    result += 'action=';
    result += action.ToString;
    result += '}';
end;

constructor TClientMessageDebugUpdateDone.Create();
begin
end;

class function TClientMessageDebugUpdateDone.ReadFrom(stream: TStream): TClientMessageDebugUpdateDone;
begin
    result := TClientMessageDebugUpdateDone.Create();
end;

procedure TClientMessageDebugUpdateDone.WriteTo(stream: TStream);
begin
    stream.WriteInt32(2);
end;

function TClientMessageDebugUpdateDone.ToString: ansistring;
begin
    result := 'DebugUpdateDone {';
    result += '}';
end;

constructor TClientMessageRequestDebugState.Create();
begin
end;

class function TClientMessageRequestDebugState.ReadFrom(stream: TStream): TClientMessageRequestDebugState;
begin
    result := TClientMessageRequestDebugState.Create();
end;

procedure TClientMessageRequestDebugState.WriteTo(stream: TStream);
begin
    stream.WriteInt32(3);
end;

function TClientMessageRequestDebugState.ToString: ansistring;
begin
    result := 'RequestDebugState {';
    result += '}';
end;

end.